File: src/container/EdgeQueue.go
package container

import (
	"github.com/HuXin0817/dots-and-boxes/src/config"
	"github.com/HuXin0817/dots-and-boxes/src/model"
	"github.com/stretchr/testify/assert"
)

type EdgeQueue struct {
	m          [model.MaxEdge]model.Edge
	front, end int
}

func (l *EdgeQueue) Clear() { l.front, l.end = 0, 0 }

func (l *EdgeQueue) Empty() bool { return l.front == l.end }

func (l *EdgeQueue) Pop() model.Edge {
	if config.DEBUG {
		assert.False(nil, l.Empty())
	}
	e := l.m[l.front]
	l.front++
	return e
}

func (l *EdgeQueue) Append(e model.Edge) {
	if config.DEBUG {
		assert.True(nil, l.end < int(model.MaxEdge))
	}
	l.m[l.end] = e
	l.end++
}

File: src/model/EdgeQueue.cpp
#include "EdgeQueue.h"

void
EdgeQueue::Clear() {
  front = 0;
  end = 0;
}

[[nodiscard]] bool
EdgeQueue::Empty() const {
  return front == end;
}

void
EdgeQueue::Append(Edge e) {
  assert(end < Edge::Max);
  m[end++] = e;
}

Edge
EdgeQueue::Pop() {
  assert(!Empty());
  auto e = m[front];
  front++;
  return e;
}

[[nodiscard]] std::vector<Edge>
EdgeQueue::Export() const {
  return {m.begin() + front, m.begin() + end};
}

[[nodiscard]] bool
EdgeQueue::Contains(Edge e) const {
  for (int i = front; i < end; i++) {
    if (static_cast<int>(m[i]) == static_cast<int>(e)) {
      return true;
    }
  }
  return false;
}
